cadcdata/cadcdata/core.py

```diff
@@ -0,0 +1,418 @@
+                        unicode_literals)
+
+import argparse
+import imp
+import logging
+import os
+import os.path
+import sys
+from StringIO import StringIO
+from datetime import datetime
+
+from cadcutils import net, util, exceptions
+from caom2.obs_reader_writer import ObservationReader, ObservationWriter
+from caom2.version import version as caom2_version
+from six.moves.urllib.parse import urlparse
+import zlib
+
+# from . import version as caom2repo_version
+from cadcdata import version
+
+__all__ = ['CadcDataClient']
+
+# TODO replace with SERVICE_URI when server supports it
+SERVICE_URL = 'www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/'
+# IVOA dateformat
+DATE_FORMAT = "%Y-%m-%dT%H:%M:%S.%f"
+DEFAULT_RESOURCE_ID = 'ivo://cadc.nrc.ca/data'
+ARCHIVE_STREAM_HTTP_HEADER = 'X-CADC-Stream'
+APP_NAME = 'cadc-data'
+
+READ_BLOCK_SIZE=8*1024
+
+logger = logging.getLogger(APP_NAME)
+
+class CadcDataClient:
+
+    """Class to access CADC archival data."""
+
+    logger = logging.getLogger('CadcDataClient')
+
+    def __init__(self, subject, resource_id=DEFAULT_RESOURCE_ID, host=None):
+        """
+        Instance of a CAOM2RepoClient
+        :param subject: the subject(user) performing the action
+        :type subject: Subject
+        :param resource_id: The identifier of the service resource (e.g 'ivo://cadc.nrc.ca/data')
+        :param host: Host server for the caom2repo service
+        """
+
+        self.resource_id = resource_id
+
+        # TODO This is just a temporary hack to be replaced with proper registry lookup functionaliy
+        resource_url = urlparse(resource_id)
+        self.host = host
+
+        agent = "cadc-data-client/{}".format(version.version)
+
+        self._repo_client = net.BaseWsClient(resource_id, subject,
+                                             agent, retry=True, host=self.host)
+        self.logger.info('Service URL: {}'.format(self._repo_client.base_url))
+
+
+    def get_file(self, archive, file_id, file=None, decompress=False,
+                 cutout=None, fhead=False, wcs=False, process_bytes=None):
+        """
+        Get a file from an archive. The entire file is delivered unless the cutout argument
+         is present specifying a cutout to extract from the file.
+        :param archive: name of the archive containing the file
+        :param file_id: the ID of the file to retrieve
+        :param file: file to save data to (file or file_name). If None, it uses the response
+        content disposition to determine the name of the destination file
+        :param cutout: perform cutout operation on the file before the result is delivered
+        :param wcs: True if the wcs is to be included with the file
+        :param process_bytes: function to be applied to the received bytes
+        :return: the data stream object
+        """
+        assert archive is not None
+        assert file_id is not None
+        resource = '/{}/{}'.format(archive, file_id)
+        params = {}
+        if fhead:
```

Comment:
To me this is clunky.  If we add or remove capabilities from the service we end up with a series of 'if' blocks.  Instead we could have a capabilities call that the dictionary builder looped over.   This would allow the argument to get_file to **kwargs type instead of named.  Then when new capabilities are added, or different ones exist for different service points, you don't need a rewrite:

for param in capabilities:
   if param in **kwargs:
      params[param] = kwargs[param]

see: http://stackoverflow.com/questions/1769403/understanding-kwargs-in-python

and others to examples of how powerful this can be.